src/main/java/io/vertx/codegen/MethodInfo.java

```diff
@@ -83,22 +84,42 @@ public String getName(Case _case) {
     return _case.format(CamelCase.INSTANCE.parse(name));
   }
 
-  public MethodKind getKind() {
+  /**
+   * @return the callback type or {@code null} when the method is not a callback method
+   */
+  public TypeInfo getCallbackType() {
     int lastParamIndex = params.size() - 1;
     if (lastParamIndex >= 0 && (returnType.isVoid() || fluent)) {
       TypeInfo lastParamType = params.get(lastParamIndex).type;
       if (lastParamType.getKind() == ClassKind.HANDLER) {
-        TypeInfo typeArg = ((ParameterizedTypeInfo) lastParamType).getArgs().get(0);
-        if (typeArg.getKind() == ClassKind.ASYNC_RESULT) {
-          return MethodKind.FUTURE;
-        } else {
-          return MethodKind.HANDLER;
-        }
+        return ((ParameterizedTypeInfo) lastParamType).getArgs().get(0);
+      }
+    }
+    return null;
+  }
+
+  public MethodKind getKind() {
+    TypeInfo callbackType = getCallbackType();
+    if (callbackType != null) {
+      if (callbackType.getKind() == ClassKind.ASYNC_RESULT && !useFutures) {
+        return MethodKind.FUTURE;
+      } else {
+        return MethodKind.HANDLER;
       }
     }
+    if (returnType.getKind() == ClassKind.FUTURE && useFutures) {
+      return MethodKind.FUTURE;
+    }
     return MethodKind.OTHER;
   }
 
+  /**
+   * @return {@code false} when the method uses the callback style
+   */
+  public boolean getUseFutures() {
```

Comment:
More than having the distinction:

* `MethodKind.FUTURE && MethodKind.getUseFutures()` for callback methods with async result
* `MethodKind.FUTURE && !MethodKind.getUseFutures()` for methods returning futures

Could we have just another `MethodKind`? For example `MethodKind.RETURN_FUTURE`. It's simpler to understand IMO and also simpler for the code generator templates.

Reply:
I think then we would rename MethodKind.FUTURE to MethodKind.CALLBACK instead and use MethodKind.FUTURE for future ?

Reply:
or something like that.

Reply:
I prefer your proposal even more!